packages/flutter/lib/src/material/material.dart

```diff
@@ -50,33 +114,333 @@ class Material extends StatelessComponent {
   }
 
   Widget build(BuildContext context) {
-    Widget contents = child;
-    if (child != null) {
+    Color backgroundColor = _getBackgroundColor(context);
+    Widget contents = config.child;
+    if (contents != null) {
       contents = new DefaultTextStyle(
-        style: textStyle ?? Theme.of(context).text.body1,
+        style: config.textStyle ?? Theme.of(context).text.body1,
         child: contents
       );
-      if (_kEdges[type] != null) {
-        contents = new ClipRRect(
-          xRadius: _kEdges[type],
-          yRadius: _kEdges[type],
-          child: contents
-        );
-      }
     }
-    return new DefaultTextStyle(
-      style: Theme.of(context).text.body1,
-      child: new AnimatedContainer(
-        curve: Curves.ease,
-        duration: kThemeChangeDuration,
-        decoration: new BoxDecoration(
-          backgroundColor: _getBackgroundColor(context),
-          borderRadius: _kEdges[type],
-          boxShadow: elevation == 0 ? null : elevationToShadow[elevation],
-          shape: type == MaterialType.circle ? Shape.circle : Shape.rectangle
-        ),
+    contents = new NotificationListener<LayoutChangedNotification>(
+      onNotification: (LayoutChangedNotification notification) {
+        _inkFeatureRenderer.currentContext.findRenderObject().markNeedsPaint();
+      },
+      child: new InkFeatures(
+        key: _inkFeatureRenderer,
+        color: backgroundColor,
         child: contents
       )
     );
+    if (config.type == MaterialType.circle) {
+      contents = new ClipOval(child: contents);
+    } else if (kMaterialEdges[config.type] != null) {
+      contents = new ClipRRect(
+        xRadius: kMaterialEdges[config.type],
+        yRadius: kMaterialEdges[config.type],
+        child: contents
+      );
+    }
+    contents = new AnimatedContainer(
+      curve: Curves.ease,
+      duration: kThemeChangeDuration,
+      decoration: new BoxDecoration(
+        backgroundColor: backgroundColor,
+        borderRadius: kMaterialEdges[config.type],
+        boxShadow: config.elevation == 0 ? null : elevationToShadow[config.elevation],
+        shape: config.type == MaterialType.circle ? Shape.circle : Shape.rectangle
+      ),
+      child: contents
+    );
+    return contents;
   }
 }
+
+const Duration _kHighlightFadeDuration = const Duration(milliseconds: 100);
+
+const double _kDefaultSplashRadius = 35.0; // logical pixels
+const int _kSplashInitialAlpha = 0x30; // 0..255
+const double _kSplashCanceledVelocity = 0.7; // logical pixels per millisecond
+const double _kSplashConfirmedVelocity = 0.7; // logical pixels per millisecond
+const double _kSplashInitialSize = 0.0; // logical pixels
+const double _kSplashUnconfirmedVelocity = 0.2; // logical pixels per millisecond
+
+class RenderInkFeatures extends RenderProxyBox implements MaterialInkController {
```

Comment:
It wasn't originally but I made it public so that you can make new `InkFeature`s (they need to pass a `RenderInkFeatures` to the `InkFeature` base class).